contxt/cli/commands/services.py

```diff
@@ -1,23 +1,219 @@
+    project = clients.contxt.get_project(from_service.stack_id)
+
+    (_, _, existing_grant) = _get_dependency_info_for_scopes(clients, from_service_id, to_service_id)
+    if not existing_grant:
+        print("Dependency does not exist between specified services")
+        return
+
+    print(f"Removing dependency between {from_service.name} -> {to_service.name}")
+    clients.deployments.remove_service_dependency(
+        organization_id=project.organization_id, service_grant=existing_grant
+    )
+    print("Success!")
+
+
+@dependencies.command()
+@click.option("--from-service-id", help="From Service ID", required=True)
+@click.option("--to-service-id", help="To Service ID", required=True)
+@click.option("--scopes", help="Command-delimited list of scope labels to add", required=True)
+@click.pass_obj
+def add_scope(clients: Clients, from_service_id: int, to_service_id: int, scopes: str) -> None:
+
+    (to_service_scopes, existing_dep_scopes, to_dep) = _get_dependency_info_for_scopes(
+        clients, from_service_id, to_service_id
+    )
+    if to_dep is None:
+        return
+
+    intended_scopes = scopes.split(",")
+    to_add = []
+    for scope in intended_scopes:
+        # check to make sure the scope is value
+        if scope not in to_service_scopes:
+            print(f'Invalid scope "{scope}" -- does not exist in to_service')
+            continue
+        if scope not in existing_dep_scopes:
+            print(f"Adding scope {scope}")
+            to_add.append(to_service_scopes[scope])
+        else:
+            print(f"{scope} already exists for dependency")
+
+    for scope in to_add:
+        r = clients.contxt.add_service_scope(to_dep, scope)
+        print(r)
+
+
+@dependencies.command()
+@click.option("--from-service-id", help="From Service ID", required=True)
+@click.option("--to-service-id", help="To Service ID", required=True)
+@click.option("--scopes", help="Command-delimited list of scope labels to add", required=True)
+@click.pass_obj
+def remove_scope(clients: Clients, from_service_id: int, to_service_id: int, scopes: str) -> None:
+
+    (to_service_scopes, existing_dep_scopes, to_dep) = _get_dependency_info_for_scopes(
+        clients, from_service_id, to_service_id
+    )
+
+    if to_dep is None:
+        return
+
+    scopes_to_remove = scopes.split(",")
+    to_remove = []
+    for scope in scopes_to_remove:
+        # check to make sure the scope actually exists
+        if scope not in to_service_scopes:
+            print(f'Invalid scope "{scope}" -- does not exist in to_service')
+            continue
+        if scope not in existing_dep_scopes:
+            print(f"Scope {scope} is not currently granted in dependency -- doing nothing")
+        else:
+            to_remove.append(to_service_scopes[scope])
+
+    for scope in to_remove:
+        clients.contxt.remove_service_scope(to_dep, scope)
+        print(f"Removed {scope}")
```

Comment:
thoughts on letting the api tell the user what is wrong rather than trying to exhaustively anticipate errors that the above loop is doing?

Reply:
I'm usually in the camp of handling things like this with "prettier" warnings specific to the command rather than some ugly exceptions coming from the API with no context of whether it's expected or not. If we didn't do this, we would see some 404 exceptions and we aren't really giving the user much context around whether it's something they need to be concerned with or not.